Add new methods to `AABB` and `Rect2` classes

mateuseap opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target other branches (e.g. `3.2`, `3.5`) if the same change was done in `master`, or is not relevant there.
PRs must not target `stable`, as that branch is updated manually.

The type of content accepted into the documentation is explained here:
https://docs.godotengine.org/en/latest/community/contributing/content_guidelines.html
-->

## What I did 

- Added the following methods to `AABB` and `Rect2` classes: `lerp()`, `cubic_interpolate()` and `cubic_interpolate_in_time()`;
- Modified the source code of following methods of `Animation` class: `_cubic_interpolate_in_time()` and `interpolate_variant()`.

#### Closes: #80627

AThousandShips commented:
In the future please be aware of when people have already said they wanted to solve an issue like this case, first ask if they are still working on it, otherwise you might cost someone a lot of lost time when you grab the issue

AThousandShips commented:
This doesn't solve the linked issue unless you also integrate these methods with the animation

mateuseap commented:
> In the future please be aware of when people have already said they wanted to solve an issue like this case, first ask if they are still working on it, otherwise you might cost someone a lot of lost time when you grab the issue

Alright, I'll pay more attention next time. For sure isn't nice to waste someone's time!

mateuseap commented:
> This doesn't solve the linked issue unless you also integrate these methods with the animation

@AThousandShips can you please give me more info of how can I do it?

AThousandShips commented:
I'll try find the relevant code later today

mateuseap commented:
Thanks, really appreciate it! I'll come back to this later as well, I'll keep looking further into the code trying to understand what else is needed to implement/modify.

TokageItLab commented:
The missing function is not the only problem, but it is the substance of the problem, which is the need to force a positive value with abs() at the end of the interpolation function.

mateuseap commented:
@AThousandShips @TokageItLab I've implemented the following methods for `AABB` and `Rect2` classes: `lerp()`, `cubic_interpolate()` and `cubic_interpolate_in_time()`. Also I started using these new methods in the source code of the following `Animation` class methods: `_cubic_interpolate_in_time()` and `interpolate_variant()`.

TokageItLab commented on core/math/aabb.h:
```diff
@@ -122,6 +122,39 @@ struct _NO_DISCARD_ AABB {
+		res.position.x = Math::lerp(res.position.x, p_to.position.x, p_weight);
+		res.position.y = Math::lerp(res.position.y, p_to.position.y, p_weight);
+		res.position.z = Math::lerp(res.position.z, p_to.position.z, p_weight);
+		res.size.x = Math::lerp(res.size.x, p_to.size.x, p_weight);
+		res.size.y = Math::lerp(res.size.y, p_to.size.y, p_weight);
+		res.size.z = Math::lerp(res.size.z, p_to.size.z, p_weight);
+		return res;
```
```suggestion
		res.size.x = Math::abs(Math::lerp(res.size.x, p_to.size.x, p_weight));
		res.size.y = Math::abs(Math::lerp(res.size.y, p_to.size.y, p_weight));
		res.size.z = Math::abs(Math::lerp(res.size.z, p_to.size.z, p_weight));
		return res;
```
Needs to apply for other parts.

mateuseap replied:
@TokageItLab should I apply the `Math::abs(Math::...` to all propreties present in the methods `lerp`, `cubic_interpolate`, `cubic_interpolate_in_time`? In case of `AABB` class it would be applied to `size` and `position` and in `Rect2` class, it would be applied to `position`, `width` and `height` props.

TokageItLab replied:
Only needs to be applied for size; negative values are allowed for position, but should not be for size.

mateuseap replied:
@TokageItLab done!

TokageItLab commented on scene/resources/animation.cpp:
```diff
@@ -5692,9 +5688,7 @@ Variant Animation::interpolate_variant(const Variant &a, const Variant &b, float
-			const Rect2 ra = a.operator Rect2();
-			const Rect2 rb = b.operator Rect2();
-			return Rect2(ra.position.lerp(rb.position, c), ra.size.lerp(rb.size, c));
+			return (a.operator Rect2()).lerp(b.operator Rect2(), c);
 		}
 		case Variant::RECT2I: {
 			const Rect2i ra = a.operator Rect2i();
```
I guess this fix would be valid for Rect2i as well. However, since it does not define the lerp in the class in the first place, and the interpolation occurs only in animation, so for now it would be fine to just apply Math::abs() to the calculation of the size in the line 5696 for avoiding error.

mateuseap replied:
Tweak applied to the code, thanks @TokageItLab!

mateuseap commented:
@TokageItLab can you take a look in this PR? I've made the requested changes

janosdebugs commented:
Hey folks, we just ran into this bug with 4.2.1.stable.official. Would it be possible to get this PR moving forward? I'd be happy to work on the solution if needed.

AThousandShips commented on core/math/aabb.h:
```diff
@@ -122,6 +122,39 @@ struct _NO_DISCARD_ AABB {
 		return position + (size * 0.5f);
 	}
 
+	_FORCE_INLINE_ AABB lerp(const AABB &p_to, const real_t p_weight) const {
```
```suggestion
	_FORCE_INLINE_ AABB lerp(const AABB &p_to, real_t p_weight) const {
```
See:
* https://github.com/godotengine/godot/pull/88467

AThousandShips commented on core/math/aabb.h:
```diff
@@ -122,6 +122,39 @@ struct _NO_DISCARD_ AABB {
+		return res;
+	}
+
+	_FORCE_INLINE_ AABB cubic_interpolate_in_time(const AABB &p_b, const AABB &p_pre_a, const AABB &p_post_b, const real_t p_weight, const real_t &p_b_t, const real_t &p_pre_a_t, const real_t &p_post_b_t) const {
```
```suggestion
	_FORCE_INLINE_ AABB cubic_interpolate_in_time(const AABB &p_b, const AABB &p_pre_a, const AABB &p_post_b, real_t p_weight, real_t p_b_t, real_t p_pre_a_t, real_t p_post_b_t) const {
```

AThousandShips commented on core/math/aabb.h:
```diff
@@ -122,6 +122,39 @@ struct _NO_DISCARD_ AABB {
+		return res;
+	}
+
+	_FORCE_INLINE_ AABB cubic_interpolate(const AABB &p_b, const AABB &p_pre_a, const AABB &p_post_b, const real_t p_weight) const {
```
```suggestion
	_FORCE_INLINE_ AABB cubic_interpolate(const AABB &p_b, const AABB &p_pre_a, const AABB &p_post_b, real_t p_weight) const {
```

AThousandShips commented on core/math/rect2.h:
```diff
@@ -51,6 +51,33 @@ struct _NO_DISCARD_ Rect2 {
 
 	_FORCE_INLINE_ Vector2 get_center() const { return position + (size * 0.5f); }
 
+	_FORCE_INLINE_ Rect2 lerp(const Rect2 &p_to, const real_t p_weight) const {
```
```suggestion
	_FORCE_INLINE_ Rect2 lerp(const Rect2 &p_to, real_t p_weight) const {
```

AThousandShips commented on core/math/rect2.h:
```diff
@@ -51,6 +51,33 @@ struct _NO_DISCARD_ Rect2 {
+		return res;
+	}
+
+	_FORCE_INLINE_ Rect2 cubic_interpolate(const Rect2 &p_b, const Rect2 &p_pre_a, const Rect2 &p_post_b, const real_t p_weight) const {
```
```suggestion
	_FORCE_INLINE_ Rect2 cubic_interpolate(const Rect2 &p_b, const Rect2 &p_pre_a, const Rect2 &p_post_b, real_t p_weight) const {
```

AThousandShips commented on core/math/rect2.h:
```diff
@@ -51,6 +51,33 @@ struct _NO_DISCARD_ Rect2 {
+		return res;
+	}
+
+	_FORCE_INLINE_ Rect2 cubic_interpolate_in_time(const Rect2 &p_b, const Rect2 &p_pre_a, const Rect2 &p_post_b, const real_t p_weight, const real_t &p_b_t, const real_t &p_pre_a_t, const real_t &p_post_b_t) const {
```
```suggestion
	_FORCE_INLINE_ Rect2 cubic_interpolate_in_time(const Rect2 &p_b, const Rect2 &p_pre_a, const Rect2 &p_post_b, real_t p_weight, real_t p_b_t, real_t p_pre_a_t, real_t p_post_b_t) const {
```

mateuseap committed:
Add new methods to AABB and Rect2 classes

- Add lerp(), cubic_interpolate() and cubic_interpolate_in_time() methods to AABB and Rect2 classes.
- Modify the animation.cpp file.
Add new methods to AABB and Rect2 classes

- Add lerp(), cubic_interpolate() and cubic_interpolate_in_time() methods to AABB and Rect2 classes.
- Modify the animation.cpp file.

mateuseap committed:
Merge branch 'master' into feat/rect2

mateuseap commented:
@AThousandShips I've modified what you've pointed out. I've also solved the conflicts.
